File: wasm/src/parser/parser.cpp
#include "parser.hpp"
#include <stdexcept>

namespace achronyme {
namespace parser {

Parser::Parser(const std::vector<Token>& tokens)
    : tokens_(tokens), current_(0) {}

std::unique_ptr<ASTNode> Parser::parse() {
    return statement();
}

// statement → let IDENTIFIER = expression | expression
std::unique_ptr<ASTNode> Parser::statement() {
    // Check for let declaration
    if (match(TokenType::LET)) {
        // let IDENTIFIER = expression
        if (!check(TokenType::IDENTIFIER)) {
            throw std::runtime_error("Expected variable name after 'let'");
        }

        std::string name = advance().lexeme;

        consume(TokenType::ASSIGN, "Expected '=' after variable name");

        auto initializer = expression();

        return std::make_unique<VariableDeclarationNode>(name, std::move(initializer));
    }

    // Otherwise, it's just an expression
    return expression();
}

// Helper methods
const Token& Parser::peek() const {
    return tokens_[current_];
}

const Token& Parser::previous() const {
    return tokens_[current_ - 1];
}

bool Parser::isAtEnd() const {
    return peek().type == TokenType::END;
}

const Token& Parser::advance() {
    if (!isAtEnd()) current_++;
    return previous();
}

bool Parser::check(TokenType type) const {
    if (isAtEnd()) return false;
    return peek().type == type;
}

bool Parser::match(TokenType type) {
    if (check(type)) {
        advance();
        return true;
    }
    return false;
}

void Parser::consume(TokenType type, const std::string& message) {
    if (check(type)) {
        advance();
        return;
    }
    throw std::runtime_error(message);
}

// Grammar rules

// expression → comparison
std::unique_ptr<ASTNode> Parser::expression() {
    return comparison();
}

// comparison → additive (('>' | '<' | '>=' | '<=' | '==' | '!=') additive)*
std::unique_ptr<ASTNode> Parser::comparison() {
    auto node = additive();

    while (match(TokenType::GT) || match(TokenType::LT) ||
           match(TokenType::GTE) || match(TokenType::LTE) ||
           match(TokenType::EQ) || match(TokenType::NEQ)) {

        TokenType op = previous().type;
        auto right = additive();

        BinaryOp binOp;
        switch (op) {
            case TokenType::GT:  binOp = BinaryOp::GT;  break;
            case TokenType::LT:  binOp = BinaryOp::LT;  break;
            case TokenType::GTE: binOp = BinaryOp::GTE; break;
            case TokenType::LTE: binOp = BinaryOp::LTE; break;
            case TokenType::EQ:  binOp = BinaryOp::EQ;  break;
            case TokenType::NEQ: binOp = BinaryOp::NEQ; break;
            default:
                throw std::runtime_error("Unknown comparison operator");
        }

        node = std::make_unique<BinaryOpNode>(binOp, std::move(node), std::move(right));
    }

    return node;
}

// additive → term (('+' | '-') term)*
std::unique_ptr<ASTNode> Parser::additive() {
    auto node = term();

    while (match(TokenType::PLUS) || match(TokenType::MINUS)) {
        TokenType op = previous().type;
        auto right = term();

        if (op == TokenType::PLUS) {
            node = std::make_unique<BinaryOpNode>(BinaryOp::ADD, std::move(node), std::move(right));
        } else {
            node = std::make_unique<BinaryOpNode>(BinaryOp::SUBTRACT, std::move(node), std::move(right));
        }
    }

    return node;
}

// term → factor (('*' | '/') factor)*
std::unique_ptr<ASTNode> Parser::term() {
    auto node = factor();

    while (match(TokenType::STAR) || match(TokenType::SLASH)) {
        TokenType op = previous().type;
        auto right = factor();

        if (op == TokenType::STAR) {
            node = std::make_unique<BinaryOpNode>(BinaryOp::MULTIPLY, std::move(node), std::move(right));
        } else {
            node = std::make_unique<BinaryOpNode>(BinaryOp::DIVIDE, std::move(node), std::move(right));
        }
    }

    return node;
}

// factor → exponent ('^' exponent)*
// Note: Right-associative! We need to handle this carefully
std::unique_ptr<ASTNode> Parser::factor() {
    auto node = exponent();

    if (match(TokenType::CARET)) {
        // Right-associative: parse the rest as a factor
        auto right = factor(); // Recursive call for right-associativity
        node = std::make_unique<BinaryOpNode>(BinaryOp::POWER, std::move(node), std::move(right));
    }

    return node;
}

// exponent → '-' exponent | primary
std::unique_ptr<ASTNode> Parser::exponent() {
    if (match(TokenType::MINUS)) {
        auto operand = exponent(); // Recursive for multiple negations: --5
        return std::make_unique<UnaryOpNode>(UnaryOp::NEGATE, std::move(operand));
    }

    return primary();
}

// primary → NUMBER 'i'? | IDENTIFIER ('(' args ')')? | '(' expression ')' 'i'? | '[' vector_or_matrix ']'
std::unique_ptr<ASTNode> Parser::primary() {
    // NUMBER with optional 'i' suffix
    if (match(TokenType::NUMBER)) {
        double value = previous().value;

        // Check for imaginary unit: 3i
        if (match(TokenType::IDENTIFIER) && previous().lexeme == "i") {
            return std::make_unique<ComplexLiteralNode>(value);  // 0 + value*i
        }

        return std::make_unique<NumberNode>(value);
    }

    // Handle identifiers (constants, function calls, lambdas, or 'i')
    if (match(TokenType::IDENTIFIER)) {
        std::string name = previous().lexeme;

        // Special case: standalone 'i' (imaginary unit)
        if (name == "i") {
            return std::make_unique<ComplexLiteralNode>(1.0);  // 0 + 1i
        }

        // Check for single-param lambda: param => expr
        if (check(TokenType::ARROW)) {
            advance(); // consume =>
            auto body = expression();
            return std::make_unique<LambdaNode>(std::vector<std::string>{name}, std::move(body));
        }

        // Check if it's a function call
        if (check(TokenType::LPAREN)) {
            return parseFunctionCall(name);
        }

        // Otherwise, it's a constant or variable reference
        return parseConstant(name);
    }

    // Parenthesized expression or multi-param lambda
    if (match(TokenType::LPAREN)) {
        // Lookahead to determine if this is a lambda or expression
        // Lambda: (x, y) => expr
        // Expression: (2 + 3)

        size_t savedPos = current_;

        // Try to parse as lambda parameters
        std::vector<std::string> params;
        bool isLambda = false;

        if (check(TokenType::IDENTIFIER)) {
            // Could be lambda
            params.push_back(advance().lexeme);

            // Check for more parameters
            while (match(TokenType::COMMA)) {
                if (!check(TokenType::IDENTIFIER)) {
                    // Not a lambda, restore position
                    current_ = savedPos;
                    goto parse_as_expression;
                }
                params.push_back(advance().lexeme);
            }

            // Must have RPAREN followed by ARROW
            if (match(TokenType::RPAREN) && check(TokenType::ARROW)) {
                isLambda = true;
                advance(); // consume =>
                auto body = expression();
                return std::make_unique<LambdaNode>(params, std::move(body));
            }
        }

        // Restore position and parse as expression
        current_ = savedPos;

parse_as_expression:
        auto node = expression();
        consume(TokenType::RPAREN, "Expected ')' after expression");

        // Check for imaginary unit: (2+3)i
        // In this case, the entire expression becomes the imaginary part
        if (match(TokenType::IDENTIFIER) && previous().lexeme == "i") {
            // We need to evaluate this at parse time if it's a simple number
            // For now, we'll handle this in a simplified way:
            // (expr)i is treated as a special case to be handled in evaluator
            // For simplicity, we'll return the node as-is and handle complex
            // construction via multiplication by 'i' in evaluator
            // Actually, let's create a UnaryOp or handle this differently

            // For now, let's just throw an error and implement the simple cases first
            throw std::runtime_error("Complex syntax (expr)i not yet fully supported. Use expr * i instead.");
        }

        return node;
    }

    // Vector or Matrix literal
    if (match(TokenType::LBRACKET)) {
        return parseVectorOrMatrix();
    }

    throw std::runtime_error("Expected expression");
}

// Parse function call: name '(' args ')'
std::unique_ptr<ASTNode> Parser::parseFunctionCall(const std::string& name) {
    consume(TokenType::LPAREN, "Expected '(' after function name");

    std::vector<std::unique_ptr<ASTNode>> args;

    // Parse arguments (if any)
    if (!check(TokenType::RPAREN)) {
        do {
            args.push_back(expression());
        } while (match(TokenType::COMMA));
    }

    consume(TokenType::RPAREN, "Expected ')' after arguments");

    return std::make_unique<FunctionCallNode>(name, std::move(args));
}

// Parse constant or variable reference
std::unique_ptr<ASTNode> Parser::parseConstant(const std::string& name) {
    // Phase 4A: Could be a constant (PI, E) or a variable (x, y)
    // Return a VariableReferenceNode and let the evaluator decide
    // (it will check variables first, then constants)
    return std::make_unique<VariableReferenceNode>(name);
}

// Parse vector or matrix starting with '['
// After consuming '[', we need to determine if this is:
//   - Vector: [expr, expr, expr]
//   - Matrix: [[expr, expr], [expr, expr]]
std::unique_ptr<ASTNode> Parser::parseVectorOrMatrix() {
    // We've already consumed the first '['

    // Check if this is a matrix (next token is '[')
    if (check(TokenType::LBRACKET)) {
        // This is a matrix: [[...], [...], ...]
        std::vector<std::vector<std::unique_ptr<ASTNode>>> rows;

        do {
            consume(TokenType::LBRACKET, "Expected '[' for matrix row");

            std::vector<std::unique_ptr<ASTNode>> row;
            if (!check(TokenType::RBRACKET)) {
                do {
                    row.push_back(expression());
                } while (match(TokenType::COMMA));
            }

            consume(TokenType::RBRACKET, "Expected ']' after matrix row");
            rows.push_back(std::move(row));

        } while (match(TokenType::COMMA));

        consume(TokenType::RBRACKET, "Expected ']' after matrix");

        // Validate that all rows have the same length
        if (!rows.empty()) {
            size_t expectedCols = rows[0].size();
            for (size_t i = 1; i < rows.size(); ++i) {
                if (rows[i].size() != expectedCols) {
                    throw std::runtime_error(
                        "Matrix rows must have the same number of elements. "
                        "Row 0 has " + std::to_string(expectedCols) +
                        " elements, but row " + std::to_string(i) +
                        " has " + std::to_string(rows[i].size()) + " elements."
                    );
                }
            }
        }

        return std::make_unique<MatrixLiteralNode>(std::move(rows));
    }

    // This is a vector: [expr, expr, expr]
    std::vector<std::unique_ptr<ASTNode>> elements;

    if (!check(TokenType::RBRACKET)) {
        do {
            elements.push_back(expression());
        } while (match(TokenType::COMMA));
    }

    consume(TokenType::RBRACKET, "Expected ']' after vector");

    return std::make_unique<VectorLiteralNode>(std::move(elements));
}

// Parse vector (called when we know it's a vector)
std::unique_ptr<ASTNode> Parser::parseVector(std::vector<std::unique_ptr<ASTNode>> firstRow) {
    return std::make_unique<VectorLiteralNode>(std::move(firstRow));
}

// Parse matrix (called when we know it's a matrix)
std::unique_ptr<ASTNode> Parser::parseMatrix(std::vector<std::unique_ptr<ASTNode>> firstRow) {
    std::vector<std::vector<std::unique_ptr<ASTNode>>> rows;
    rows.push_back(std::move(firstRow));

    // Continue parsing remaining rows
    while (match(TokenType::COMMA)) {
        consume(TokenType::LBRACKET, "Expected '[' for matrix row");

        std::vector<std::unique_ptr<ASTNode>> row;
        do {
            row.push_back(expression());
        } while (match(TokenType::COMMA));

        consume(TokenType::RBRACKET, "Expected ']' after matrix row");
        rows.push_back(std::move(row));
    }

    return std::make_unique<MatrixLiteralNode>(std::move(rows));
}

} // namespace parser
} // namespace achronyme

File: crates/achronyme-parser/src/parser.rs
use crate::ast::{AstNode, BinaryOp, UnaryOp};
use crate::token::Token;

pub struct Parser {
    tokens: Vec<Token>,
    current: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self { tokens, current: 0 }
    }

    // Helper methods
    fn peek(&self) -> &Token {
        &self.tokens[self.current]
    }

    fn previous(&self) -> &Token {
        &self.tokens[self.current - 1]
    }

    fn is_at_end(&self) -> bool {
        matches!(self.peek(), Token::Eof)
    }

    fn advance(&mut self) -> &Token {
        if !self.is_at_end() {
            self.current += 1;
        }
        self.previous()
    }

    fn check(&self, expected: &Token) -> bool {
        if self.is_at_end() {
            return false;
        }
        std::mem::discriminant(self.peek()) == std::mem::discriminant(expected)
    }

    fn match_token(&mut self, expected: &Token) -> bool {
        if self.check(expected) {
            self.advance();
            true
        } else {
            false
        }
    }

    fn consume(&mut self, expected: &Token, message: &str) -> Result<&Token, String> {
        if self.check(expected) {
            Ok(self.advance())
        } else {
            Err(message.to_string())
        }
    }

    // Main parse entry point
    pub fn parse(&mut self) -> Result<AstNode, String> {
        self.statement()
    }

    // statement → let IDENTIFIER = expression | expression
    fn statement(&mut self) -> Result<AstNode, String> {
        // Check for let declaration
        if self.match_token(&Token::Let) {
            // let IDENTIFIER = expression
            if let Token::Identifier(name) = self.peek().clone() {
                self.advance();
                self.consume(&Token::Assign, "Expected '=' after variable name")?;
                let initializer = self.expression()?;
                return Ok(AstNode::VariableDecl {
                    name,
                    initializer: Box::new(initializer),
                });
            } else {
                return Err("Expected variable name after 'let'".to_string());
            }
        }

        // Otherwise, it's just an expression
        self.expression()
    }

    // expression → comparison
    fn expression(&mut self) -> Result<AstNode, String> {
        self.comparison()
    }

    // comparison → additive (('>' | '<' | '>=' | '<=' | '==' | '!=') additive)*
    fn comparison(&mut self) -> Result<AstNode, String> {
        let mut node = self.additive()?;

        while self.match_token(&Token::Gt)
            || self.match_token(&Token::Lt)
            || self.match_token(&Token::Gte)
            || self.match_token(&Token::Lte)
            || self.match_token(&Token::Eq)
            || self.match_token(&Token::Neq)
        {
            let op = match self.previous() {
                Token::Gt => BinaryOp::Gt,
                Token::Lt => BinaryOp::Lt,
                Token::Gte => BinaryOp::Gte,
                Token::Lte => BinaryOp::Lte,
                Token::Eq => BinaryOp::Eq,
                Token::Neq => BinaryOp::Neq,
                _ => return Err("Unknown comparison operator".to_string()),
            };

            let right = self.additive()?;
            node = AstNode::BinaryOp {
                op,
                left: Box::new(node),
                right: Box::new(right),
            };
        }

        Ok(node)
    }

    // additive → term (('+' | '-') term)*
    fn additive(&mut self) -> Result<AstNode, String> {
        let mut node = self.term()?;

        while self.match_token(&Token::Plus) || self.match_token(&Token::Minus) {
            let op = match self.previous() {
                Token::Plus => BinaryOp::Add,
                Token::Minus => BinaryOp::Subtract,
                _ => unreachable!(),
            };

            let right = self.term()?;
            node = AstNode::BinaryOp {
                op,
                left: Box::new(node),
                right: Box::new(right),
            };
        }

        Ok(node)
    }

    // term → factor (('*' | '/' | '%') factor)*
    fn term(&mut self) -> Result<AstNode, String> {
        let mut node = self.factor()?;

        while self.match_token(&Token::Star)
            || self.match_token(&Token::Slash)
            || self.match_token(&Token::Modulo)
        {
            let op = match self.previous() {
                Token::Star => BinaryOp::Multiply,
                Token::Slash => BinaryOp::Divide,
                Token::Modulo => BinaryOp::Modulo,
                _ => unreachable!(),
            };

            let right = self.factor()?;
            node = AstNode::BinaryOp {
                op,
                left: Box::new(node),
                right: Box::new(right),
            };
        }

        Ok(node)
    }

    // factor → exponent ('^' exponent)*
    // Note: Right-associative! We need to handle this carefully
    fn factor(&mut self) -> Result<AstNode, String> {
        let mut node = self.exponent()?;

        if self.match_token(&Token::Caret) {
            // Right-associative: parse the rest as a factor
            let right = self.factor()?; // Recursive call for right-associativity
            node = AstNode::BinaryOp {
                op: BinaryOp::Power,
                left: Box::new(node),
                right: Box::new(right),
            };
        }

        Ok(node)
    }

    // exponent → '-' exponent | primary
    fn exponent(&mut self) -> Result<AstNode, String> {
        if self.match_token(&Token::Minus) {
            // Recursive for multiple negations: --5
            let operand = self.exponent()?;
            return Ok(AstNode::UnaryOp {
                op: UnaryOp::Negate,
                operand: Box::new(operand),
            });
        }

        self.primary()
    }

    // primary → NUMBER 'i'? | IDENTIFIER ('(' args ')')? | '(' expression ')' 'i'? | '[' vector_or_matrix ']'
    fn primary(&mut self) -> Result<AstNode, String> {
        // NUMBER with optional 'i' suffix
        if let Token::Number(value) = self.peek() {
            let value = *value;
            self.advance();

            // Check for imaginary unit: 3i
            if let Token::Identifier(name) = self.peek() {
                if name == "i" {
                    self.advance();
                    return Ok(AstNode::ComplexLiteral {
                        re: 0.0,
                        im: value,
                    });
                }
            }

            return Ok(AstNode::Number(value));
        }

        // Handle identifiers (constants, function calls, lambdas, or 'i')
        if let Token::Identifier(name) = self.peek().clone() {
            self.advance();

            // Special case: standalone 'i' (imaginary unit)
            if name == "i" {
                return Ok(AstNode::ComplexLiteral { re: 0.0, im: 1.0 });
            }

            // Check for single-param lambda: param => expr
            if self.check(&Token::Arrow) {
                self.advance(); // consume =>
                let body = self.expression()?;
                return Ok(AstNode::Lambda {
                    params: vec![name],
                    body: Box::new(body),
                });
            }

            // Check if it's a function call
            if self.check(&Token::LParen) {
                return self.parse_function_call(name);
            }

            // Otherwise, it's a constant or variable reference
            return Ok(AstNode::VariableRef(name));
        }

        // Parenthesized expression or multi-param lambda
        if self.match_token(&Token::LParen) {
            // Lookahead to determine if this is a lambda or expression
            // Lambda: (x, y) => expr
            // Expression: (2 + 3)

            let saved_pos = self.current;

            // Try to parse as lambda parameters
            let mut params = Vec::new();
            let mut is_lambda = false;

            if let Token::Identifier(name) = self.peek().clone() {
                params.push(name);
                self.advance();

                // Check for more parameters
                while self.match_token(&Token::Comma) {
                    if let Token::Identifier(name) = self.peek().clone() {
                        params.push(name);
                        self.advance();
                    } else {
                        // Not a lambda, restore position
                        self.current = saved_pos;
                        params.clear();
                        break;
                    }
                }

                // Must have RPAREN followed by ARROW
                if self.match_token(&Token::RParen) && self.check(&Token::Arrow) {
                    is_lambda = true;
                    self.advance(); // consume =>
                    let body = self.expression()?;
                    return Ok(AstNode::Lambda {
                        params,
                        body: Box::new(body),
                    });
                }
            }

            // Restore position and parse as expression
            if !is_lambda {
                self.current = saved_pos;
            }

            let node = self.expression()?;
            self.consume(&Token::RParen, "Expected ')' after expression")?;

            // Check for imaginary unit: (2+3)i
            // For simplicity, we'll throw an error for now (as C++ does)
            if let Token::Identifier(name) = self.peek() {
                if name == "i" {
                    return Err(
                        "Complex syntax (expr)i not yet fully supported. Use expr * i instead."
                            .to_string(),
                    );
                }
            }

            return Ok(node);
        }

        // Vector or Matrix literal
        if self.match_token(&Token::LBracket) {
            return self.parse_vector_or_matrix();
        }

        Err("Expected expression".to_string())
    }

    // Parse function call: name '(' args ')'
    fn parse_function_call(&mut self, name: String) -> Result<AstNode, String> {
        self.consume(&Token::LParen, "Expected '(' after function name")?;

        let mut args = Vec::new();

        // Parse arguments (if any)
        if !self.check(&Token::RParen) {
            loop {
                args.push(self.expression()?);
                if !self.match_token(&Token::Comma) {
                    break;
                }
            }
        }

        self.consume(&Token::RParen, "Expected ')' after arguments")?;

        Ok(AstNode::FunctionCall { name, args })
    }

    // Parse vector or matrix starting with '['
    // After consuming '[', we need to determine if this is:
    //   - Vector: [expr, expr, expr]
    //   - Matrix: [[expr, expr], [expr, expr]]
    fn parse_vector_or_matrix(&mut self) -> Result<AstNode, String> {
        // We've already consumed the first '['

        // Check if this is a matrix (next token is '[')
        if self.check(&Token::LBracket) {
            // This is a matrix: [[...], [...], ...]
            let mut rows = Vec::new();

            loop {
                self.consume(&Token::LBracket, "Expected '[' for matrix row")?;

                let mut row = Vec::new();
                if !self.check(&Token::RBracket) {
                    loop {
                        row.push(self.expression()?);
                        if !self.match_token(&Token::Comma) {
                            break;
                        }
                    }
                }

                self.consume(&Token::RBracket, "Expected ']' after matrix row")?;
                rows.push(row);

                if !self.match_token(&Token::Comma) {
                    break;
                }
            }

            self.consume(&Token::RBracket, "Expected ']' after matrix")?;

            // Validate that all rows have the same length
            if !rows.is_empty() {
                let expected_cols = rows[0].len();
                for (i, row) in rows.iter().enumerate() {
                    if row.len() != expected_cols {
                        return Err(format!(
                            "Matrix rows must have the same number of elements. \
                             Row 0 has {} elements, but row {} has {} elements.",
                            expected_cols,
                            i,
                            row.len()
                        ));
                    }
                }
            }

            return Ok(AstNode::MatrixLiteral(rows));
        }

        // This is a vector: [expr, expr, expr]
        let mut elements = Vec::new();

        if !self.check(&Token::RBracket) {
            loop {
                elements.push(self.expression()?);
                if !self.match_token(&Token::Comma) {
                    break;
                }
            }
        }

        self.consume(&Token::RBracket, "Expected ']' after vector")?;

        Ok(AstNode::VectorLiteral(elements))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;

    fn parse(source: &str) -> Result<AstNode, String> {
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize()?;
        let mut parser = Parser::new(tokens);
        parser.parse()
    }

    #[test]
    fn test_simple_number() {
        let ast = parse("42").unwrap();
        assert_eq!(ast, AstNode::Number(42.0));
    }

    #[test]
    fn test_addition() {
        let ast = parse("2 + 3").unwrap();
        match ast {
            AstNode::BinaryOp { op, .. } => assert_eq!(op, BinaryOp::Add),
            _ => panic!("Expected BinaryOp"),
        }
    }

    #[test]
    fn test_power_right_associative() {
        // 2^3^4 should be 2^(3^4), not (2^3)^4
        let ast = parse("2 ^ 3 ^ 4").unwrap();
        match ast {
            AstNode::BinaryOp {
                op: BinaryOp::Power,
                left,
                right,
            } => {
                assert_eq!(*left, AstNode::Number(2.0));
                match *right {
                    AstNode::BinaryOp {
                        op: BinaryOp::Power,
                        ..
                    } => {} // Correct: right side is another power operation
                    _ => panic!("Power should be right-associative"),
                }
            }
            _ => panic!("Expected power operation"),
        }
    }

    #[test]
    fn test_precedence() {
        // 2 + 3 * 4 should be 2 + (3 * 4)
        let ast = parse("2 + 3 * 4").unwrap();
        match ast {
            AstNode::BinaryOp {
                op: BinaryOp::Add,
                left,
                right,
            } => {
                assert_eq!(*left, AstNode::Number(2.0));
                match *right {
                    AstNode::BinaryOp {
                        op: BinaryOp::Multiply,
                        ..
                    } => {} // Correct
                    _ => panic!("Multiplication should have higher precedence"),
                }
            }
            _ => panic!("Expected addition at top level"),
        }
    }

    #[test]
    fn test_vector() {
        let ast = parse("[1, 2, 3]").unwrap();
        match ast {
            AstNode::VectorLiteral(elements) => assert_eq!(elements.len(), 3),
            _ => panic!("Expected vector literal"),
        }
    }

    #[test]
    fn test_function_call() {
        let ast = parse("sin(PI)").unwrap();
        match ast {
            AstNode::FunctionCall { name, args } => {
                assert_eq!(name, "sin");
                assert_eq!(args.len(), 1);
            }
            _ => panic!("Expected function call"),
        }
    }

    #[test]
    fn test_let_statement() {
        let ast = parse("let x = 5").unwrap();
        match ast {
            AstNode::VariableDecl { name, .. } => {
                assert_eq!(name, "x");
            }
            _ => panic!("Expected variable declaration"),
        }
    }

    #[test]
    fn test_lambda() {
        let ast = parse("x => x * 2").unwrap();
        match ast {
            AstNode::Lambda { params, .. } => {
                assert_eq!(params.len(), 1);
                assert_eq!(params[0], "x");
            }
            _ => panic!("Expected lambda"),
        }
    }

    #[test]
    fn test_comparison() {
        let ast = parse("x > 5").unwrap();
        match ast {
            AstNode::BinaryOp { op, .. } => assert_eq!(op, BinaryOp::Gt),
            _ => panic!("Expected comparison"),
        }
    }
}
